golos.publication/golos.publication.cpp

```diff
@@ -334,31 +334,58 @@ void publication::unvote(name voter, structures::mssgid message_id) {
-            INLINE_ACTION_SENDER(token, payment) (config::token_name, {_self, config::code_name}, {_self, user, quantity, memo});
-        } 
-        else {
-            INLINE_ACTION_SENDER(token, payment) (config::token_name, {_self, config::code_name}, {_self, _self, quantity, memo});
-        }
-    }
-    else if(static_cast<payment_t>(mode) == payment_t::VESTING) {
-        if (golos::vesting::balance_exist(config::vesting_name, user, quantity.symbol.code())) {
-            INLINE_ACTION_SENDER(token, transfer) (config::token_name, {_self, config::code_name},
-            {_self, config::vesting_name, quantity, std::string(config::send_prefix + name{user}.to_string()  + "; " + memo)});
-        } 
-        else {
-            INLINE_ACTION_SENDER(token, payment) (config::token_name, {_self, config::code_name}, {_self, _self, quantity, memo});
+    auto set_payout = [&](payment_t mode) {
+        auto check_balance = [&](payment_t mode, name acc, symbol sym) -> bool {
+            if (mode == payment_t::TOKEN)
+                return token::balance_exist(config::token_name, acc, sym.code());
+            else
+                return golos::vesting::balance_exist(config::vesting_name, acc, sym.code());
+        };
+        
+        for (auto recipient_obj : recipients) {
+            if (recipient_obj.quantity.amount >= 0) {
+                if (check_balance(mode, recipient_obj.to, recipient_obj.quantity.symbol)) {
+                    if (recipient_obj.quantity.amount)
+                        recipients_payout.push_back(recipient_obj);
+                } else {
+                    if (recipient_obj.quantity.amount)
+                        publish_payout.amount += recipient_obj.quantity.amount;
+                }
+            } else {
+                wrong_quantity.push_back(recipient_obj.to);
+            }
         }
-    }
-    else
+    };
+
+    if (static_cast<payment_t>(mode) == payment_t::TOKEN) {
+        set_payout(payment_t::TOKEN); 
+        INLINE_ACTION_SENDER(token, bulkpayment) (config::token_name, {_self, config::code_name}, {_self, recipients_payout});
+        print("\npay token");
+    } else if (static_cast<payment_t>(mode) == payment_t::VESTING) {
+        set_payout(payment_t::VESTING); 
+        INLINE_ACTION_SENDER(token, bulktransfer) (config::token_name, {_self, config::code_name}, {_self, recipients_payout});
+        print("\npay vesting");
+    } else
         eosio_assert(false, "publication::payto: unknown kind of payment");
+    
+    if (publish_payout.amount) {
+        print("\npay publish before");
+        INLINE_ACTION_SENDER(token, payment) (config::token_name, {_self, config::code_name},
+                {_self, _self, publish_payout, "pay reward of post to publish account"});
+        print("\npay publish after");
+    }
+
+    eosio_assert(!wrong_quantity.size(), "LOGIC ERROR! publication::payto: quantity.amount < 0");
```

Comment:
Why do you split reaction on negative balance from check in line 356? If you place `eosio_assert(recipient_obj.quantity.amount >= 0, ...` in 356 line code will become cleaner. And you can print recipient and balance at that point.